proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+    block will not be executed. Since we cannot control behaviour of `init` block execution, we restrict it for inline classes.
+    
+    Unfortunately, it's not enough, because `init` blocks can be emulated via factory methods:
+    ```kotlin
+    inline class Positive private constructor(val value: Int) {
+        companion object {
+            fun create(x: Int) {
+                assert(x > 0) "Value isn't positive: x"
+                return Positive(x)  
+            }  
+        }
+    }
+  
+    fun foo(p: Positive) {}
+    ```
+    
+    Again, method `foo` have type `int` from Java POV, so we can indirectly create values of type `Positive` 
+    even with the presence of private constructor.
+    
+    To make behaviour more predictable and consistent with Java, we demand public primary constructor and restrict `init` blocks.
+
+### Other restrictions
+
+The following restrictions are related to the usages of inline classes:
+
+- Referential equality (`===`) is prohibited for inline classes
+- vararg of inline class type is prohibited
+```
+inline class Foo(val s: String)
+
+fun test(vararg foos: Foo) { ... } // should be an error  
+```
+
+## Java interoperability  
+
+Each inline class has its own wrapper, which is boxed/unboxed by the same rules as for primitive types.
+Basically, rule for boxing can be formulated as follows: inline class is boxed when it is used as another type.
+Unboxed inline class is used when value is statically known to be inline class.
+
+Examples:
+```kotlin
+interface I
+
+inline class Foo(val i: Int) : I
+
+fun asInline(f: Foo) {}
+fun <T> asGeneric(x: T) {}
+fun asInterface(i: I) {}
+fun asNullable(i: Foo?) {}
+
+fun <T> id(x: T): T = x
+
+fun test(f: Foo) {
+    asInline(f)
+    asGeneric(f) // boxing
+    asInterface(f) // boxing
+    asNullable(f) // boxing
+    
+    val c = id(f) // boxing/unboxing, c is unboxed
+}
+```
+
+Since boxing doesn't have side effects as is, it's possible to reuse various optimizations that are done for primitive types.
+
+### Type mapping on JVM
+
+Depending on the underlying type of inline class and its declaration site, inline class type can be mapped to the underlying type or to the 
+boxed type.  
+
+| Underlying Type \ Declaration Site | Not-null | Nullable | Generic |
+| --------------------------------------------- | --------------- | ---------------- | ---------------- |
+| **Reference type** | Underlying type | Underlying type | Boxed type |
+| **Nullable reference type** | Underlying type  | Boxed type | Boxed type |
+| **Primitive type** | Underlying type  | Boxed type | Boxed type |
+| **Nullable primitive type** | Underlying type  | Boxed type | Boxed type |
+
+For example, from this table it follows that nullable inline class which is based on reference type is mapped to the underlying type:
+```kotlin
+inline class Name(val s: String)
+
+fun foo(n: Name?) { ... } // `Name?` here is mapped to String
```

Comment:
Shouldn't it be `String?`

Reply:
No, here we know that `null` can be associated only with the type `Name?`, because underlying type of the inline class `Name` is a non-null reference type.

Reply:
This is similar to what we have for `IntArray`/`LongArray`/...: `IntArray?` is just an array of primitive types